File: test/proto/scala/Smoke.scala
// Smoke test for the Scala test provider port. Prints summary stats that must
// match the canonical TS output documented in PROVIDER.md.
//
// Run from the repo root (Scala 3 toolchain required):
//   cd test/proto/scala && scalac Provider.scala Smoke.scala -d out \
//     && (cd /home/user/struct && scala -cp test/proto/scala/out Smoke)

import scala.collection.immutable.TreeMap

object Smoke {

  def main(args: Array[String]): Unit = {
    val path = if (args.length > 0) Some(args(0)) else Some("build/test/test.json")
    val prov = TestProvider.load(path)

    val fns = prov.functions()
    println("functions: " + fns.mkString(", "))

    var total = 0
    var expectKinds = TreeMap.empty[String, Int]
    var inputKinds = TreeMap.empty[String, Int]
    for (fn <- fns) {
      for (e <- prov.entries(fn)) {
        total += 1
        val ek = e.expect.kind.toString.toLowerCase
        val ik = e.input.kind.toString.toLowerCase
        expectKinds = expectKinds.updated(ek, expectKinds.getOrElse(ek, 0) + 1)
        inputKinds = inputKinds.updated(ik, inputKinds.getOrElse(ik, 0) + 1)
      }
    }

    println("total entries: " + total)
    println("expect kinds: " + joinCounts(expectKinds))
    println("input kinds: " + joinCounts(inputKinds))

    val e = prov.entries("getpath", Some("basic")).head
    println(
      "getpath/basic[0]: id=" + e.id.orNull +
        ", doc=" + e.doc +
        ", input.kind=" + e.input.kind.toString.toLowerCase +
        ", expect.kind=" + e.expect.kind.toString.toLowerCase +
        ", expect.value=" + e.expect.value.map(TestMatch.stringify).orNull
    )
  }

  private def joinCounts(counts: TreeMap[String, Int]): String =
    counts.map { case (k, v) => s"$k=$v" }.mkString(", ")
}

File: test/proto/java/Smoke.java
// Smoke test for the Java test provider port. Prints summary stats that must
// match the canonical TS output documented in PROVIDER.md.
//
// Run from the repo root:
//   cd test/proto/java && javac *.java && (cd /home/user/struct && java -cp test/proto/java Smoke)

import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class Smoke {

  public static void main(String[] args) {
    String path = args.length > 0 ? args[0] : "build/test/test.json";
    Provider prov = Provider.load(path);

    List<String> fns = prov.functions();
    System.out.println("functions: " + String.join(", ", fns));

    int total = 0;
    Map<String, Integer> expectKinds = new TreeMap<>();
    Map<String, Integer> inputKinds = new TreeMap<>();
    for (String fn : fns) {
      for (Provider.Entry e : prov.entries(fn, null)) {
        total++;
        String ek = e.expect.kind.name().toLowerCase();
        String ik = e.input.kind.name().toLowerCase();
        expectKinds.merge(ek, 1, Integer::sum);
        inputKinds.merge(ik, 1, Integer::sum);
      }
    }

    System.out.println("total entries: " + total);
    System.out.println("expect kinds: " + joinCounts(expectKinds));
    System.out.println("input kinds: " + joinCounts(inputKinds));

    Provider.Entry e = prov.entries("getpath", "basic").get(0);
    System.out.println(
        "getpath/basic[0]: id="
            + e.id
            + ", doc="
            + e.doc
            + ", input.kind="
            + e.input.kind.name().toLowerCase()
            + ", expect.kind="
            + e.expect.kind.name().toLowerCase()
            + ", expect.value="
            + Provider.stringify(e.expect.value));
  }

  private static String joinCounts(Map<String, Integer> counts) {
    StringBuilder sb = new StringBuilder();
    boolean first = true;
    for (Map.Entry<String, Integer> c : counts.entrySet()) {
      if (!first) {
        sb.append(", ");
      }
      first = false;
      sb.append(c.getKey()).append('=').append(c.getValue());
    }
    return sb.toString();
  }
}
